File: 433-minimum-genetic-mutation/minimum-genetic-mutation.java
class Solution {
    public int minMutation(String startGene, String endGene, String[] bank) {
        Set<String> geneBank = new HashSet<>(Arrays.asList(bank));
        if(!geneBank.contains(endGene)) {
            return -1;
        }

        char[] mutations = {'A','C','G','T'};

        int mutationCount = 0;
        Queue<String> queue = new LinkedList<>();
        queue.add(startGene);
        while(!queue.isEmpty()) {
            int levelSize = queue.size();
            for(int i = 0; i < levelSize; i++) {
                String gene = queue.poll();
                if(gene.equals(endGene)) {
                    return mutationCount;
                }

                StringBuilder currGene = new StringBuilder(gene);
                for(int index = 0; index < currGene.length(); index++) {
                    char original = currGene.charAt(index);
                    for(char mutation : mutations) {
                        currGene.setCharAt(index, mutation);
                        String nextGene = currGene.toString();
                        if(geneBank.contains(nextGene)) {
                            queue.add(nextGene);
                            geneBank.remove(nextGene);
                        }
                    }
                    currGene.setCharAt(index, original);
                }
            }
            mutationCount++;
        }

        return -1;
    }
}

File: 433-minimum-genetic-mutation/minimum-genetic-mutation.cpp
class Solution {
public:
    int minMutation(string startGene, string target, vector<string>& bank) 
    {
        unordered_set<string> set(bank.begin(), bank.end());

        // can't go to the target if its not in the bank
        if(!set.count(target))
        return -1;

        queue<string> q;
        q.push(startGene);

        int steps = 0;
        vector<char> mutations = {'A', 'C', 'G', 'T'};
        while(!q.empty())
        {
            int level = q.size();
            while(level--)
            {
                string currGene = q.front();
                q.pop();

                if(currGene == target)
                return steps;

                for(int i = 0; i < currGene.length(); i++)
                {
                    int org = currGene[i];
                    for(char ch = 'A'; ch <= 'Z'; ch++)
                    {
                        if(ch == org)
                        continue;

                        // change the character
                        currGene[i] = ch;

                        // check whether this mutataed gene exists in the bank
                        if(set.count(currGene))
                        {
                            q.push(currGene);
                            set.erase(currGene);
                        }
                    }
                    // change the character at this index back to original
                    currGene[i] = org;
                }
            }
            steps++;
        }
        return -1;
    }
};